source/opt/instrument_pass.cpp

```diff
@@ -1008,78 +1008,122 @@ bool InstrumentPass::InstrumentFunction(Function* func, uint32_t stage_idx,
-      stage != SpvExecutionModelGeometry &&
-      stage != SpvExecutionModelGLCompute &&
-      stage != SpvExecutionModelTessellationControl &&
-      stage != SpvExecutionModelTessellationEvaluation &&
-      stage != SpvExecutionModelTaskNV && stage != SpvExecutionModelMeshNV &&
-      stage != SpvExecutionModelRayGenerationNV &&
-      stage != SpvExecutionModelIntersectionNV &&
-      stage != SpvExecutionModelAnyHitNV &&
-      stage != SpvExecutionModelClosestHitNV &&
-      stage != SpvExecutionModelMissNV &&
-      stage != SpvExecutionModelCallableNV &&
-      stage != SpvExecutionModelTaskEXT && stage != SpvExecutionModelMeshEXT) {
-    if (consumer()) {
-      std::string message = "Stage not supported by instrumentation";
-      consumer()(SPV_MSG_ERROR, 0, {0, 0, 0}, message.c_str());
-    }
-    return false;
   }
-  // Add together the roots of all entry points
-  std::queue<uint32_t> roots;
+
+  // Next, collect together the roots and stage indices of all entry points
+  bool mixed_stages = false;
+  std::map<uint32_t, std::queue<uint32_t>> stage_to_roots;
   for (auto& e : get_module()->entry_points()) {
-    roots.push(e.GetSingleWordInOperand(kEntryPointFunctionIdInIdx));
+    uint32_t current_stage =
+        e.GetSingleWordInOperand(kEntryPointExecutionModelInIdx);
+    uint32_t current_root =
+        e.GetSingleWordInOperand(kEntryPointFunctionIdInIdx);
+    stage_to_roots[current_stage].push(current_root);
+    if (current_stage != stage) mixed_stages = true;
   }
-  bool modified = InstProcessCallTreeFromRoots(pfn, &roots, stage);
+
+  if (mixed_stages) {
+    // For each stage type, collect a set of function calls
+    std::map<uint32_t, std::unordered_set<uint32_t>> stage_to_functions;
+    for (auto& str_it : stage_to_roots) {
+      uint32_t current_stage = str_it.first;
+      auto roots_to_process = str_it.second;
+      while (!roots_to_process.empty()) {
+        uint32_t root = roots_to_process.front();
+        roots_to_process.pop();
+        context()->CollectCallTreeFromRoots(root,
+                                            &stage_to_functions[current_stage]);
+      }
+    }
+
+    // Check for any potential intersections between function sets.
+    // TODO(greg-lunarg): Handle mixed stages. Technically, a shader module
+    // can contain entry points with different execution models, although
+    // such modules will likely be rare as GLSL and HLSL are geared toward
+    // one model per module. In such cases, we will need to clone any functions
+    // which are in the call trees of entrypoints with differing execution
+    // models.
+
+    // EDIT(natevm): this scenario is actually common in ray tracing pipelines,
+    // as one module might contain a raygen, miss, and closesthit all in one.
+    // As a temporary workaround, here we allow mixed shader stages so long
+    // as entry points do not share common function calls.
```

Comment:
No need to keep the older comments in the code.  Just say

```
// Check if there is a function that might be called with different shader stages.  The instrumentation is
// specific to the shader stage, and that function cannot be instrumented two different ways.  We do not
// expect this to be common because ...
```

Note that one case that might come up every once in a while will be the OpKill function.  See WrapOpKill for the details.  You might want to special case that function because it could cause you problems.